### DISCUSSION TITLE
The publishing API

### DISCUSSION THREAD

**Discussion by radu-matei** [bytecodealliance/SIG-Registries#1] [General]
This is the thread we will use to asynchronously discuss the publishing API for the registry.

Proposal PR: https://github.com/bytecodealliance/SIG-Registries/pull/10

**Comment by lann** [comment]
As introduced in today's meeting: https://hackmd.io/IegsPA7IQDOXJgPg7hGByA

**Comment by rvolosatovs** [comment]
A few comments on the proposal as discussed during the May 4 meeting:
- The proposed API does not address crypto agility in any way and forces usage of SHA256 - could that be a problem, especially in the future? Some organizations may require usage of one or the other hashing algorithm.
- Digest representation format - any reason not to use https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity or something similar (as long as it's URL-safe)?
- The suggested format does not store any metadata of the contents, e.g. length/size of the contents - should some metric be stored along the digest to provide better integrity guarantees?

cc @npmccallum

**Reply by lann** [reply]
> The proposed API does not address crypto agility in any way and forces usage of SHA256 - could that be a problem, especially in the future? Some organizations may require usage of one or the other hashing algorithm.

I don't think that cryptographic agility should necessarily be a default approach. Famously, [many JWT libraries were vulnerable](https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/) due to improperly-implemented agility. I would be happy to discuss the tradeoffs of agility in this design, but I would like to root that discussion in concrete use cases.

**Reply by lann** [reply]
> Digest representation format - any reason not to use https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity or something similar (as long as it's URL-safe)?

Nope, I just selected the simplest encoding at hand.

**Reply by lann** [reply]
> The suggested format does not store any metadata of the contents, e.g. length/size of the contents - should some metric be stored along the digest to provide better integrity guarantees?

I can't think of any specific cryptographic need to authenticate the content length in this design. The only thing that comes to mind would be a [Length extension attack](https://en.wikipedia.org/wiki/Length_extension_attack) which applies to certain naive MAC implementations, but I don't believe it is applicable here.

**Reply by carlsverre** [reply]
Re: the standard attack on cryptographic agility regarding JWTs

This doesn't apply to content digests. It applies to signatures. For content digests there isn't an equivalent attack to the ones mentioned in the blog post. The goal here is to represent content digests in a flexible format that allows us to enforce new requirements on new libraries without needing to re-hash all previously stored information (or at least do it incrementally).

As for cryptographic agility in signatures, there is another discussion focused on that: #7

**Reply by lann** [reply]
I am just generally wary of any avoidable complexity in cryptographic system design. You're right that the JWT agility problem doesn't specifically apply here, but it is representative of the kind of complexity I worry about.

**Reply by guybedford** [reply]
Upgrading the hash seems doable, and a retrospective bulk operation wouldn't be impossible. This seems an unavoidable problem, so at least making a future upgrade possible seems the best path? Also is there a reason not to go with 384 or higher to be safe?

The publish API looks like a really great start to me though with exactly the right level of detail with comprehensive content guarantees.
